Fails on JDK 16 due to caffeine version

I got the following exception when I tried to run a sample modular program with JDK 16:

```
Exception in thread "main" java.lang.NoClassDefFoundError: sun/misc/Unsafe
	at com.github.benmanes.caffeine@2.8.0/com.github.benmanes.caffeine.base.UnsafeAccess.load(UnsafeAccess.java:68)
	at com.github.benmanes.caffeine@2.8.0/com.github.benmanes.caffeine.base.UnsafeAccess.<clinit>(UnsafeAccess.java:42)
	at com.github.benmanes.caffeine@2.8.0/com.github.benmanes.caffeine.cache.BLCHeader$DrainStatusRef.<clinit>(BoundedLocalCache.java:3912)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:78)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:480)
	at com.github.benmanes.caffeine@2.8.0/com.github.benmanes.caffeine.cache.LocalCacheFactory.newBoundedLocalCache(LocalCacheFactory.java:98)
	at com.github.benmanes.caffeine@2.8.0/com.github.benmanes.caffeine.cache.BoundedLocalCache$BoundedLocalManualCache.<init>(BoundedLocalCache.java:3392)
	at com.github.benmanes.caffeine@2.8.0/com.github.benmanes.caffeine.cache.BoundedLocalCache$BoundedLocalLoadingCache.<init>(BoundedLocalCache.java:3716)
	at com.github.benmanes.caffeine@2.8.0/com.github.benmanes.caffeine.cache.Caffeine.build(Caffeine.java:1014)
	at org.jdbi.v3.core@3.18.1/org.jdbi.v3.core.statement.CachingSqlParser.<init>(CachingSqlParser.java:30)
	at org.jdbi.v3.core@3.18.1/org.jdbi.v3.core.statement.CachingSqlParser.<init>(CachingSqlParser.java:25)
	at org.jdbi.v3.core@3.18.1/org.jdbi.v3.core.statement.ColonPrefixSqlParser.<init>(ColonPrefixSqlParser.java:41)
	at org.jdbi.v3.core@3.18.1/org.jdbi.v3.core.statement.SqlStatements.<init>(SqlStatements.java:53)
	at org.jdbi.v3.core@3.18.1/org.jdbi.v3.core.config.ConfigRegistry.lambda$null$3(ConfigRegistry.java:100)
	at org.jdbi.v3.core@3.18.1/org.jdbi.v3.core.internal.exceptions.Unchecked.lambda$function$4(Unchecked.java:76)
	at org.jdbi.v3.core@3.18.1/org.jdbi.v3.core.config.ConfigRegistry.get(ConfigRegistry.java:78)
	at org.jdbi.v3.core@3.18.1/org.jdbi.v3.core.config.ConfigRegistry.<init>(ConfigRegistry.java:47)
	at org.jdbi.v3.core@3.18.1/org.jdbi.v3.core.Jdbi.<init>(Jdbi.java:58)
	at org.jdbi.v3.core@3.18.1/org.jdbi.v3.core.Jdbi.create(Jdbi.java:104)
	at org.jdbi.v3.core@3.18.1/org.jdbi.v3.core.Jdbi.create(Jdbi.java:88)
	at td.nextgen/td.Main.main(Main.java:55)
Caused by: java.lang.ClassNotFoundException: sun.misc.Unsafe
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:636)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:182)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:519)
	... 24 more
```

Bumping the caffeine cache library version to 3.0.1 solves the problem. Reference: https://github.com/ben-manes/caffeine/issues/273

--- a/README.md
+++ b/README.md
@@ -15,10 +15,13 @@ Jdbi with it.
 
 Jdbi 3 requires Java 8 or better to run. Jdbi 3 requires Java 11 or better to compile.
 
-We run CI tests against Java 8, 11, 14, and 15.  We recommend running the latest GA JDK.
+We run CI tests against Java 8, 11, 14, and 15. We recommend running the latest GA JDK.
 
-At this point Java 8 is considered deprecated.  While Jdbi does not (yet) have a specific
+At this point Java 8 is considered deprecated. While Jdbi does not (yet) have a specific
 date to drop support, please chart your path forward to a supported JDK!
+
+NOTE: to run on Java 8, you may need to manage the `caffeine` dependency back to the
+latest 2.x release. 3.x is necessary for newer JDKs but does not run on 8.
 
 ## Builds
 

--- a/RELEASE_NOTES.md
+++ b/RELEASE_NOTES.md
@@ -1,6 +1,7 @@
 # 3.19.0
   - Simple slf4j SqlLogger implementation to debug all statements executed
   - new RollbackOnlyTransactionHandler rolls back everything for testing
+  - upgrade caffeine dep to 3.0.1 for jdk16 (NOTE: jdk8 users will need to manage it back to 2.x)
 
 # 3.18.1
   - Comments like -- and // now recognized and discarded from SQL, thanks @rherrmann!

--- a/docs/src/adoc/index.adoc
+++ b/docs/src/adoc/index.adoc
@@ -144,6 +144,15 @@ module you want to use:
 Jdbi provides several other modules, which enhance the core API with additional
 features.
 
+=== Java Compatibility
+
+`Jdbi` runs on all Java versions 8 or later. 11 or later is requied to build.
+
+[WARNING]
+Java 8 support is considered deprecated and will be maintained best-effort for now,
+but will be going away soon! In order to run on 8, you might need to dependency-manage your
+`caffeine` version back to 2.x. 3.x is required to run on newer JDKs, but will not run on 8.
+
 === Modules
 
 [glossary]

--- a/pom.xml
+++ b/pom.xml
@@ -84,6 +84,7 @@
         <basepom.test.timeout>240</basepom.test.timeout>
 
         <dep.antlr.version>4.9</dep.antlr.version>
+        <dep.caffeine.version>3.0.1</dep.caffeine.version>
         <dep.dokka.version>1.4.30</dep.dokka.version>
         <dep.freebuilder.version>2.6.1</dep.freebuilder.version>
         <dep.immutables.version>2.7.1</dep.immutables.version>
@@ -253,7 +254,7 @@
             <dependency>
                 <groupId>com.github.ben-manes.caffeine</groupId>
                 <artifactId>caffeine</artifactId>
-                <version>2.8.0</version>
+                <version>${dep.caffeine.version}</version>
                 <exclusions>
                     <exclusion>
                         <artifactId>error_prone_annotations</artifactId>

--- a/src/build/ci-test.sh
+++ b/src/build/ci-test.sh
@@ -19,7 +19,7 @@ if [ "$JDK_VERSION" = "8" ]; then
         PATH=${OLD_PATH}
         JAVA_HOME=${OLD_JAVA_HOME}
         export JAVA_HOME PATH
-        exec mvn -DargLine= surefire:test
+        exec mvn -Ddep.caffeine.version=2.8.0 -DargLine= surefire:test
     fi
 fi
 

